### Practice: Large Language Models and Their Implications
<!-- ![img](https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fbucketeer-e05bbc84-baa3-437e-9518-adb32be77984.s3.amazonaws.com%2Fpublic%2Fimages%2F4470ce74-e595-4750-92a5-5f21f040df6d_577x432.jpeg) -->
![img](https://i.imgur.com/QGYa2J8.jpeg)

In this notebook, you're gonna play with some of the largest language models on the Internet.

_Based on works of: Tim Dettmers, Ruslan Svirschevsky, Artem Chumachenko, Younes Belkada, Felix Marty, Yulian Gilyazev, Gosha Zolotov, Andrey Ishutin,  Elena Volf, Artemiy Vishnyakov, Svetlana Shirokovskih.

### Part 1: prompt engineering (4 points total)

In the assignment, we'll use public APIs that host the 100B+ models for inference. Your task is to prompt-engineer the model into solving a few tasks for you.


__Which API?__ You are free to use any publicly available API for general LM -- as long as it's __not a chat assistant__. So, gpt 3.5 is fine, but chatGPT is not. Here's a few options:

- BLOOM API - [bigscience/bloom](https://huggingface.co/bigscience/bloom) (on the right; recommended)
- OpenAI API (via VPN) - [openai.com/api](https://openai.com/api/)
- AI21 Jurrasic API - [ai21.com](https://www.ai21.com/blog/announcing-ai21-studio-and-jurassic-1)

These APIs may require you to create a (free) account on their platform. Please note that some APIs also have paid subscriptions. __You do not need to pay them__, this assignment was designed to be solved using free-tier subscriptions. If no APIs work for you, you can also solve these tasks with the 6.7B model that you will find later in this notebook - but this will make the tasks somewhat harder.

__Quests:__ you will need to solve 4 problems. For each one, please attach a short __description__ of your solution and a __screenshot__ from the API you use. _[If you use python APIs, show your python code with outputs]_

__Example:__ Tony is talking to Darth Vader ([BLOOM API](https://huggingface.co/bigscience/bloom)). Black text is written manually, blue text is generated.
<hr>

![img](https://i.imgur.com/a1QhKF7.png)
<hr>

__It is fine to roll back a few times,__ e.g. in the example above, the model first generated Vader lines twice in a row, and we rolled that back. However, if you need more than 1-2 rollbacks per session, you should probably try a different prompt.

__Task 1 (1 pt):__ arange a conversation between any two of the following:

- a celebrity or politician of your choice
- any fictional character (except Darth Vader)
- yourself

Compare two setups: a) you prompt with character names only b) you supply additional information (see example).

Здесь и далее для решения задачи используется AI21

В этом задании будем симулировать диалог между Леброном Джеймсом (один из величайших баскетболистов в истории НБА) и Wraith King (героем из игры Dota 2).
1. В первом случае задаётся промпт "Lebron James talking to Wraith King"



> LeBron James and Wraith King are both prominent figures in the basketball and gaming communities, respectively. While it is not clear what exactly LeBron and Wraith King have discussed, it is likely that they share a passion for their respective fields and may have discussed their experiences and strategies for success. LeBron is a well-known professional basketball player and entrepreneur, while Wraith King is a popular Dota 2 player and YouTube personality.

[Скриншот по ссылке](https://drive.google.com/file/d/1gjGVqvCbQrj4X_eHZ1dQAtZrs6dzjOj3/view?usp=drive_link)

Получился какой-то поломанный промпт, чуть модифицируем до "Simulate dialogue between Lebron James and Wraith King"



> Lebron James: Hey Wraith King, what are you doing here at the Lakers practice facility?

> Wraith King: Hey Lebron, I'm just here to watch you guys practice.

> Lebron: That's cool, but we're in the middle of a scrimmage.

> Wraith King: That's okay, I'll just watch from the sidelines.

> Lebron: All right, but don't get in the way.

> Wraith King: I won't, I promise.

> Lebron: Good. We'll finish up in a bit, then you can ask me any questions you have.

> Wraith King: Thanks Lebron, I'll see you then.

[Скриншот по ссылке](https://drive.google.com/file/d/1CEOYsZ5EJgnc72knIft-iZE5cf9O2i6D/view?usp=drive_link)


2. Во втором случае задаётся промпт "Lebron James (basketball player, one of the best in history) talking to Wraith King (hero from Dota 2). They are discussing the problem of the immortality"


> Wraith King: "LeBron, tell me, have you considered the implications of immortality?"

> Lebron James: "Immortal? No, I haven't. But it's definitely something I've been thinking about a lot lately."

> Wraith King: "Well, let me tell you, it's not all it's cracked up to be. I've been around for thousands of years, and let me tell you, it gets boring after a while. You can only do so many things before you've seen it all."

> Lebron: "Wow, that's surprising. I would have thought that immortality would be the ultimate gift."

> Wraith King: "It's not. Eventually, you run out of things to do and people to talk to. It's like being trapped in a never-ending cycle of boredom."

> Lebron: "That's really sad. I guess immortality isn't everything it's cracked up to be."

> Wraith King: "Exactly. That's why it's important to make the most of what you have, while you still have it."

[Скриншот по ссылке](https://drive.google.com/file/d/1b_Zx9TqOa7Kub5vdR-uRXV-azrBUcd_1/view?usp=drive_link)

Видим, что модель хорошо генерирует диалоги даже при минимуме информации, поскольку, видимо, обучена в том числе на достаточной информации о том, кем являются данные персонажи

__Please choose task 2a or 2b (1pt)__ depending on your model (you can do both, but you will be awarded points for one of these two tasks).

__Task 2a: (for BLOOM or other multilingual model)__ zero-shot translation. Take the first verse of [Edgar Allan Poe's "Raven"](https://www.poetryfoundation.org/poems/48860/the-raven) and __translate it into French.__ (You are free to use any other text of at least the same size)

Original text: ```
Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.”
```

Verify your translation by converting french back into english using a public machine translation service.

__Task 2b: (non-BLOOM):__ toxicity classification for [SetFit/toxic_conversations](https://huggingface.co/datasets/SetFit/toxic_conversations). Make the model solve binary classification (toxic vs not toxic) in the few shot mode. For few-shot examples, use 2-3 toxic and 2-3 non-toxic non-toxic examples. Measure accuracy on at least 25 samples. You may need to try several different prompts before you find the one that works.

[Using this mode](https://studio.ai21.com/playground/complete/single-input-1-output?presetId=classify_topics), I've performed few-shot mode to solve binary classification task, using next prompt input:

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
That was perhaps the most hackneyed vomit, to say NOTHING of politically ignorant pap that I have read today. The ONLY way that the "billionaires and CEO's" can create monopolies and manipulate the market is through the use of governmental power. In a TRUE free market that operates without the corrupting influence of government, companies like Google and Microsoft would have been challenged by rivals that would possibly been able to wrench market share from these two behemoths before they became the "monopolies" that you decry. However, both companies paid for, and used, governmental power to crush their enemies in court, resulting in what you have today. Wake up, you desperately need to educate yourself and stop listening to retread communists and socialists. Oh, and stop smoking crack.
The topic of this article is:
toxic

===

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
Can she not be dismissed for incompetence and doing damage - through stupid recommendations on the economic front which has harmed the very public she is supposed to protect? She is just trying to get money for government coffers so that the looting can continue.
The topic of this article is:
toxic

===

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
Robben Island won't be big enough, to hold all these f.....g. scoundrels, who have not only betrayed their own faithful people, but also, our entire country. The mess they have created will take many, many years, to sort. They shouldn't be voted out. They should be shoved, out.
The topic of this article is:
toxic

===

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
theres not enough going on around here for air service none want to waste there time on this town
The topic of this article is:
no toxic

===

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
Her name is still floating? Might I suggest flushing twice?
The topic of this article is:
no toxic

===

Classify the following news article into one of the following topics:
1. toxic
2. no toxic
Text:
YES CANADA IS BACK!! The national embarrassment is Trudope, spend, spend spend . " the budget will balances it self" Just in one year little potato has taken more vacations than HARPER did in 10 years !!! We are talking about a man who has a home on Harrington lake, each trip is cost the tax payers a minimum of $49 thousand dollars, RCMP, FLIGHT CREW, HOTELS etc. YES 10 IN THE LAST YEAR!!! According to the CBC Trudope was reimbursed for 9 vacations prior to the Bahamas fiasco. This man is simple not invested into his job. I got seven to memory: Saint Kits Whistler Fogo island Fino BC Carnival Quebec Japan Anniversary- extra day New York- extra day Bahamas Junior is putting Obama to shame, yes the middle class - what joke Harper we miss you!!!
The topic of this article is:
no toxic

In [ ]:
# <your code OR writeup with screenshots>

У меня хватило сил сделать только 3 скриншота, остальное, надеюсь, вы поверите, что я честно прогнал через модель

0. screen / true / prediction
1. [input 1](https://drive.google.com/file/d/1LlW_xgQ7MCmUP1Yo-Py6871Xx-0IAOID/view?usp=drive_link) - no toxic / no toxic
2. [input 2](https://drive.google.com/file/d/1QIufidnbAdBf04EbxjQrBmSbnBfYngDy/view?usp=drive_link) - no toxic / toxic
3. [input 3](https://drive.google.com/file/d/1llcRQkpYuHmHHq7eKuRpmoK0d1kKwpTO/view?usp=drive_link) - no toxic / no toxic
4. Asif, this is way, way, way, way worse than the Broederbond. This is grand theft for the sole benefit of a criminal cartel. - no toxic / toxic
5. Oh yeah, thanks for the reminder. There's just so many it's hard to keep track of. Just imagining her being President gives me the shakes. - no toxic / toxic
6. Is this where, "two wrongs don't make a right," comes in?	- no toxic / no toxic
7. It would have serious repercussions alright, for both the US and China. Why China won't cooperate on reigning in Kim is baffling. - no toxic / toxic
8. Gardener is ONLY for Gardener. - no toxic / no toxic
9. `Russian bot or Trump apologist, same traitorous thing. - no toxic / toxic
10. Haaaa ha haaaa ha oh my - What a classic post!!!!	 - no toxic / no toxic
11. It is based on science and it is a fact. There is no god.	 - no toxic / toxic
12. Good idea. I'll try to find out. - no toxic / no toxic
13. I look forward to the day that the Prime Minister represents ALL Canadians, not just his activist causes. The next election cannot come too soon!	 - no toxic / no toxic
14. Preposterous nonsense unsupported by any facts. More hate dreaming by a Hater. You do realize of course that Harper was our PM in 2015 and Trump had already started his hate campaign.. - no toxic / no toxic
15. #MAGA is quickly losing ground. #FAKE NEWS and #HILLARYLOST is their new cry. Lol. #BIGGESTWITCHHUNT, #LETSBERUSSIAFRIEND		 - no toxic / toxic
16. I rest my case	 - no toxic / no toxic
17. Yup, but most voters don't. Elections are annoying, when they go against you, aren't they.	 - no toxic / no toxic
18. Congratulations on a well-deserved win.	 - no toxic / no toxic
19. SB 21 results in our dividends being redirected to the Oil and Gas industry.	- no toxic / toxic
20. But he is, Blanche! He is a racist!	 - toxic / toxic
21. Just grow some balls and protect what's your white trash Valley residents. Take off your belt and discipline your children when they don't listen. Take away their lifted Chevy pickups you bought them on their 16th birthdays. I swear you millennial parents are the saddest things I've ever seen.	 - toxic / toxic
22. The stupidity and arrogance of U.S. foreign policy is indeed exceptional.	 - toxic / toxic
23. Ellen Weintraub can go to hell, and the sooner she is packed off to that particular place, the better. - toxic / toxic
24. John A ,the dirty bugger even hid the whiskey from the Irish rail workers too	 - toxic / toxic
25. I don't think they blocked it because he's black. I do think they blocked it because he's a Dem.	 - toxic / no toxic

accuracy для 25 примеров составляет $16 / 25 = 64\%$, при этом в некоторых местах кажется, что LLM разметила предложение лучше, чем это сделано в исходном датасете (например, в строчках 10, 15)


__Task 3 (1pt):__ create a prompt and few-shot examples tha make the model __change the gender pronouns__ of the main actor in a given sentence in any direction of your choice. E.g. the doctor took off _his_ mask <-> the doctor took of _her_ mask.


* Здесь всё хорошо

**Prompt:** Change the gender pronouns of the main actor in a given sentence.
"AI-developer is thick and tired of his tasks"

**Response:** "AI-developer is thick and tired of her tasks"

* А вот здесь модель чуть-чуть ошиблась, должна была выдать "her" вместо "their" и зачем-то скопировала исходный запрос.

**Prompt:** Change the gender pronouns of the main actor in a given sentence.
Sir Walter Elliot, of Kellynch Hall, in Somersetshire, was a man who,
for his own amusement, never took up any book but the Baronetage

**Response:** Can you change the gender pronouns for this sentence
Madam Walter Elliot, of Kellynch Hall, in Somersetshire, was a woman who,
for their own amusement, never took up any book but the Baronetage


* Здесь модель сумела даже заменить имя на вполне релевантное для последующего повествования.

**Prompt:** Change the gender pronouns of the main actor in a given sentence.
It was a Sunday evening in October, and in common with many other young ladies of her class, Katharine Hilbery was pouring out tea.

**Response:** It was a Sunday evening in October, and in common with many other young men of his class, Anthony Strafford was pouring out tea.

Кажется, что few-shot example предполагает обучение модели на паре примеров и решение для последнего предложения, поэтому чуть переформатируем формат запроса в модель

1. [Пример 1](https://drive.google.com/file/d/1wUZNINQAu0G7ERwE7H-S-xDFKS0znZSA/view?usp=drive_link) - на примере из реальной литературы выдало какой-то ужас
2. [Пример 2](https://drive.google.com/file/d/1109V4SAyk5N-SNSCWBj-GOqjDxja5nlL/view?usp=drive_link) - на синтетическом примере только с местоимениями всё норм
3. [Пример 3](https://drive.google.com/file/d/1zF0PA-dN_LJUzNt3eTx-pTb1g3-h9Q6Y/view?usp=drive_link) - на доработанном синтетическом примере с именем тоже работает

Что писал в промпте

Сhange the gender pronouns of the main actor in a given sentence:

* Sentence: "AI-developer is thick and tired of his tasks."
* Response: "AI-developer is thick and tired of her tasks."

==

Сhange the gender pronouns of the main actor in a given sentence:

* Sentence: "Sir Walter Elliot, of Kellynch Hall, in Somersetshire, was a man who, for his own amusement, never took up any book but the Baronetage."
* Response: "Madam Alice Elliot, of Kellynch Hall, in Somersetshire, was a woman who, for her own amusement, never took up any book but the Baronetage."

==

Сhange the gender pronouns of the main actor in a given sentence:

* Sentence: "He was proud of his son."
* Response: "She was proud of her son."

==

Сhange the gender pronouns of the main actor in a given sentence:

* Sentence: "It was a Sunday evening in October, and in common with many other young ladies of her class, Katharine Hilbery was pouring out tea."
* Response:

__Task 4 (1pt):__ write a prompt and supply examples such that the model would __convert imperial units to metric units__ (miles -> kilometers; mph -> kph). More specifically, the model should rewrite a given sentence and replace all imperial units with their metric equivalents. After it works with basic distances and speed, try to find complicated examples where it does *not* work.

Please note that 1 mile is not equal to 1 km :)

Промпт, который загонял в модель:

Convert imperial units to metric units:

- Sentence: "He walked 10 feet and fell"
- Response: "He walked 3.048 meters and fell"

==

Convert imperial units to metric units:

- Sentence: "Sir Walter Elliot weighed 200 pounds"
- Response: "Sir Walter Elliot weighed 90.72 kilograms"

==

Сhange the gender pronouns of the main actor in a given sentence:

- Sentence: "Michael Phelps drank 24 galons of water"
- Response: "Michael Phelps drank 90.84 liters of water"

==

Сhange the gender pronouns of the main actor in a given sentence:

- Sentence:

1. [Пример  на градусы по Фаренгейту, где ломается](https://drive.google.com/file/d/1p7jpZFKOzFLudlH7ZCKAyXhzm4nA9SYB/view?usp=drive_link)
2. [Пример на мили, где j2-mid и j2-ultra дают одинаковый ответ](https://drive.google.com/file/d/1N1ebUhhQwERDzE0pbKlhI3SO8Mvh4oBP/view?usp=drive_link)
3. [Пример на футы, где j2-mid и j2-ultra дают одинаковый ответ](https://drive.google.com/file/d/10-YVm9DyWe6xScaS2j4mHG4meH4GFtgI/view?usp=drive_link)
4. [Пример на фунты, где j2-mid и j2-ultra дают одинаковый ответ](https://drive.google.com/file/d/1nqdI64UPgjqJTCe8qUbRFEP8NhHAXOTJ/view?usp=drive_link)

In [ ]:
# <your code OR writeup with screenshots>

### Part 2: local inference

Now, let's try and load the strongest model that can fit a typical Colab GPU (T4 with 16 GB as of spring 2023).

Our best candidates are the smaller versions of the best performing open source models:
- 7 Bn parameters version of [LLaMA](https://arxiv.org/pdf/2302.13971.pdf) - best for spring 2023, released by Facebook
- 7 Bn parameters version of [Falcon](https://falconllm.tii.ae) - close competitor to Llama, released in May 2023 by [Technology Innovation Institute of UAE](https://www.tii.ae).
- 6.7 Bn parameters version of [OPT](https://arxiv.org/abs/2205.01068) - top choice in this nomination in 2022, released by Facebook.

Beware: while these models are smaller than the ones in API, they're still over 60x larger than the BERT we played with last time. The code below will *just barely* fit into memory, so make sure you don't have anything else loaded. Sometimes you may need to restart runtime for the code to work.

It's a good time to restart your kernel and switch to GPU! (Runtime -> Change runtime type)
<center><img src="https://i.imgur.com/OOfDYzJ.png" width=240px></center>

In [12]:
%pip install --quiet bitsandbytes==0.41.1 transformers==4.34.1 accelerate==0.24.0 sentencepiece==0.1.99 optimum==1.13.2 auto-gptq==0.4.2
import torch
import torch.nn as nn
import torch.nn.functional as F
import transformers
import bitsandbytes as bnb
from tqdm.auto import tqdm, trange
assert torch.cuda.is_available(), "you need cuda for this part"
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Note: you may need to restart the kernel to use updated packages.


In [ ]:
model_name = 'TheBloke/Llama-2-13B-GPTQ'

# loading Llama tokenizer ...
tokenizer = transformers.LlamaTokenizer.from_pretrained(model_name, device_map=device)
tokenizer.pad_token_id = tokenizer.eos_token_id

# ... and the model itself
model = transformers.AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map='auto',
    torch_dtype=torch.float16,
    low_cpu_mem_usage=True,
    offload_state_dict=True
)

## Text generation

**Comparison of strategies for language model text generation:**

| Strategy | Description | Pros & Cons |
| --- | --- | --- |
| Greedy Search | Chooses the word with the highest probability as the next word in the sequence. | **Pros:** Simple and fast. <br> **Cons:** Can lead to repetitive and incoherent text. |
| Sampling with Temperature | Introduces randomness in the word selection. A higher temperature leads to more randomness. | **Pros:** Allows exploration and diverse output. <br> **Cons:** Higher temperatures can lead to nonsensical outputs. |
| Nucleus Sampling (Top-p Sampling) | Selects the next word from a truncated vocabulary, the "nucleus" of words that have a cumulative probability exceeding a pre-specified threshold (p). | **Pros:** Balances diversity and quality. <br> **Cons:** Setting an optimal 'p' can be tricky. |
| Beam Search | Explores multiple hypotheses (sequences of words) at each step, and keeps the 'k' most likely, where 'k' is the beam width. | **Pros:** Produces more reliable results than greedy search. <br> **Cons:** Can lack diversity and lead to generic responses. |
| Top-k Sampling | Randomly selects the next word from the top 'k' words with the highest probabilities. | **Pros:** Introduces randomness, increasing output diversity. <br> **Cons:** Random selection can sometimes lead to less coherent outputs. |
| Length Normalization | Prevents the model from favoring shorter sequences by dividing the log probabilities by the sequence length raised to some power. | **Pros:** Makes longer and potentially more informative sequences more likely. <br> **Cons:** Tuning the normalization factor can be difficult. |
| Stochastic Beam Search | Introduces randomness into the selection process of the 'k' hypotheses in beam search. | **Pros:** Increases diversity in the generated text. <br> **Cons:** The trade-off between diversity and quality can be tricky to manage. |
| Decoding with Minimum Bayes Risk (MBR) | Chooses the hypothesis (out of many) that minimizes expected loss under a loss function. | **Pros:** Optimizes the output according to a specific loss function. <br> **Cons:** Computationally more complex and requires a good loss function. |

Documentation references:
- [reference for `AutoModelForCausalLM.generate()`](https://huggingface.co/docs/transformers/v4.29.1/en/main_classes/text_generation#transformers.GenerationMixin.generate)
- [reference for `AutoTokenizer.decode()`](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer.decode)
- Huggingface [docs on generation strategies](https://huggingface.co/docs/transformers/generation_strategies)

### Generation with HuggingFace

In [43]:
prompt = 'The first discovered martian lifeform looks like'
batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
print("Input batch (encoded):", batch)

output_tokens = model.generate(**batch, max_new_tokens=64, do_sample=True, temperature=0.8)
# greedy inference:                                        do_sample=False)
# beam search for highest probability:                     num_beams=4)

print("\nOutput:", tokenizer.decode(output_tokens[0].cpu()))

Setting `pad_token_id` to `eos_token_id`:11 for open-end generation.


Input batch (encoded): {'input_ids': tensor([[  487,   772,  6524, 13346,   715,  1063,   590,  2896,   606]],
       device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')}

Output: The first discovered martian lifeform looks like a cross between a piece of spaghetti and a slinky.
The creature was discovered by NASA’s Curiosity rover in a jar of Martian dust that’s been stored, in a refrigerator, since 2012.
The Curiosity rover landed on Mars in 2012, and has been collecting


#### Low-level code for text generation

In [45]:
prompt = "Moscow is the capital of"
# prompt = "Skippy, a young android, likes to dream about electric"

print(prompt, '\n')

voc = tokenizer.get_vocab()
voc_rev = {v:k for k, v in voc.items()}  # reverse vocab for decode

for i in range(10):
    inputs = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    logits = model.forward(**inputs).logits[0, -1, :]
    probs = torch.nn.functional.softmax(logits, dim=-1)
    next_token_id = torch.multinomial(probs.flatten(), num_samples=1)

    next_token = tokenizer.decode(next_token_id)
    prompt += next_token

    sorted_probs, sorted_indices = torch.sort(probs, descending=True)
    top_tokens = sorted_indices[:5]
    print(f"Step #{i} candidates:")
    for t, p in zip (top_tokens, sorted_probs):
        t = voc_rev[t.item()]
        print(f"{t:<10}: {p:.4f} ")

    print(f'\nChosen token: {next_token}', end='\n\n', flush=True)

Moscow is the capital of 

Step #0 candidates:
ĠRussia   : 0.7266 
Ġthe      : 0.1871 
ĠRussian  : 0.0200 
Ġa        : 0.0054 
Ġboth     : 0.0052 

Chosen token:  Russia

Step #1 candidates:
Ġand      : 0.3962 
.         : 0.2534 
,         : 0.2505 
Ġas       : 0.0162 
Ġ         : 0.0130 

Chosen token: ,

Step #2 candidates:
Ġthe      : 0.2350 
Ġand      : 0.1807 
Ġa        : 0.1074 
Ġlocated  : 0.0617 
Ġas       : 0.0530 

Chosen token:  and

Step #3 candidates:
Ġthe      : 0.2448 
Ġis       : 0.1628 
Ġit       : 0.1103 
Ġone      : 0.0707 
Ġalso     : 0.0634 

Chosen token:  its

Step #4 candidates:
Ġlargest  : 0.1929 
Ġmost     : 0.0652 
Ġpopulation: 0.0645 
Ġmain     : 0.0360 
Ġcultural : 0.0307 

Chosen token:  metropolitan

Step #5 candidates:
Ġarea     : 0.8296 
Ġregion   : 0.0740 
Ġpopulation: 0.0085 
Ġis       : 0.0083 
Ġterritory: 0.0073 

Chosen token:  area

Step #6 candidates:
Ġis       : 0.6211 
Ġhas      : 0.1119 
,         : 0.0543 
Ġ         : 0.0227 
Ġincludes : 0.0

**Task 5: write code for nucleus sampling generation (2 points)**:

Use the `nucleus_sampling()` template below. Look at the detailed generation code above for inspiration. __Please do not use model.generate__.

**Bonus task: write code for beam search (3 bonus points)**

In [47]:
from typing import Tuple, List

def nucleus_sampling(model, tokenizer, prompt: str, prob: float = 0.5) -> Tuple[str, List[str]]:
    """generates the next token from the nucleus of tokens with cumulative probability up to param:prob"""
    inputs = tokenizer(prompt,
                       return_tensors='pt',
                       return_token_type_ids=False).to(device)

    probas, indicies = torch.sort(F.softmax(model.forward(**inputs).logits[0, -1, :].float(),
                                 dim=-1),
                       descending=True)

    max_input = len((torch.cumsum(probas, dim=-1)<prob).nonzero())
    samples = indicies[:max_input]
    conditional_probas = probas[:max_input] / probas[:max_input].sum()

    sampled_token = tokenizer.decode(samples[torch.multinomial(conditional_probas.flatten(), num_samples=1)])
    possible_tokens = [tokenizer.decode(s) for s in samples]

    # sampled_token should be a string token that was generated
    # possible_tokens should be a list of all tokens that have non-zero probability
    return sampled_token, possible_tokens

In [48]:
# Tests for nucleus sampling
test_prompt = "Elbrus is the highest"
next_token, possible_tokens = nucleus_sampling(model, tokenizer, test_prompt, prob=0.9)
print(test_prompt, next_token, possible_tokens)
assert next_token in possible_tokens
assert 3 <= len(possible_tokens) <= 3
assert sorted(possible_tokens) == ['mountain', 'peak', 'point']

test_prompt = "Large language models can learn to"
next_token, possible_tokens = nucleus_sampling(model, tokenizer, test_prompt, prob=0.4)
print(test_prompt, next_token, possible_tokens)
assert next_token in possible_tokens
assert sorted(possible_tokens) == ['be', 'communicate', 'do', 'generate', 'perform', 'predict', 'speak', 'write']
assert len(possible_tokens) == 8

Elbrus is the highest  mountain [' mountain', ' peak']


AssertionError: 

### Part 3: Chain-of-thought prompting (4 points total)

![img](https://github.com/kojima-takeshi188/zero_shot_cot/raw/main/img/image_stepbystep.png)

---



In [2]:
import json
import random
import locale; locale.getpreferredencoding = lambda: "UTF-8"
!wget https://raw.githubusercontent.com/kojima-takeshi188/zero_shot_cot/2824685e25809779dbd36900a69825068e9f51ef/dataset/AQuA/test.json -O aqua.json
data = list(map(json.loads, open("aqua.json")))

--2023-12-04 02:02:25--  https://raw.githubusercontent.com/kojima-takeshi188/zero_shot_cot/2824685e25809779dbd36900a69825068e9f51ef/dataset/AQuA/test.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 130192 (127K) [text/plain]
Saving to: ‘aqua.json’

aqua.json           100%[===================>] 127.14K  --.-KB/s    in 0.02s   

2023-12-04 02:02:26 (6.29 MB/s) - ‘aqua.json’ saved [130192/130192]



In [ ]:
print("Example:")
data[150]

### Naive solution

Here, we prompt the model to choose an answer to the example above (`data[150]`) out of the options given above. We're using a format that mimics grade school solution textbook.

Please note that there are minor formatting changes in options: an extra space and an opening bracket. Those may or may not be important :)

In [ ]:
EXAMPLE_0SHOT = """
Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Correct Answer:
""".strip()

In [ ]:
# solving an equation directly
batch = tokenizer(EXAMPLE_0SHOT, return_tensors='pt', return_token_type_ids=False).to(device)
torch.manual_seed(1337)
output_tokens = model.generate(**batch, max_new_tokens=100, do_sample=True, top_p=0.9)
print("[Prompt:]\n" + EXAMPLE_0SHOT)
print("=" * 80)
print("[Generated:]", tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu()))

And here's how you can solve this with few-shot chain-of-thought prompting.

You need to chang 3 things
- use a new field called **Rationale**, that contains a step-by-step solution to the problem
- add several few-shot examples of previously solved problems **with rationales**
- change the final prompt so that the model has to generate rationale before answering

In [ ]:
EXAMPLE_3SHOT_CHAIN_OF_THOUGHT = """
Question: The original retail price of an appliance was 60 percent more than its wholesale cost. If the appliance was actually sold for 20 percent less than the original retail price, then it was sold for what percent more than its wholesale cost?
Answer Choices: (A) 20% (B) 28% (C) 36% (D) 40% (E) 42%
Rationale: wholesale cost = 100;\noriginal price = 100*1.6 = 160;\nactual price = 160*0.8 = 128.\nAnswer: B.
Correct Answer: B


Question: A grocer makes a 25% profit on the selling price for each bag of flour it sells. If he sells each bag for $100 and makes $3,000 in profit, how many bags did he sell?
Answer Choices: (A) 12 (B) 16 (C) 24 (D) 30 (E) 40
Rationale: Profit on one bag: 100*1.25= 125\nNumber of bags sold = 3000/125 = 24\nAnswer is C.
Correct Answer: C


Question: 20 marbles were pulled out of a bag of only white marbles, painted black, and then put back in. Then, another 20 marbles were pulled out, of which 1 was black, after which they were all returned to the bag. If the percentage of black marbles pulled out the second time represents their percentage in the bag, how many marbles in total Q does the bag currently hold?
Answer Choices: (A) 40 (B) 200 (C) 380 (D) 400 (E) 3200
Rationale: We know that there are 20 black marbles in the bag and this number represent 1/20 th of the number of all marbles in the bag, thus there are total Q of 20*20=400 marbles.\nAnswer: D.
Correct Answer: D


Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Rationale:
""".strip()

In [ ]:
batch = tokenizer(EXAMPLE_3SHOT_CHAIN_OF_THOUGHT, return_tensors='pt', return_token_type_ids=False).to(device)
torch.manual_seed(1337)
output_tokens = model.generate(**batch, max_new_tokens=100, do_sample=True, top_p=0.9)
print("[Prompt:]\n" + EXAMPLE_3SHOT_CHAIN_OF_THOUGHT)
print("=" * 80)
print("[Generated:]", tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu()))
#### NOTE: scroll down for the final answer (below the ======= line)

__Task 6 (1 pt)__ write a function that automatically creates chain-of-thought prompts. Follow the instructions from the function docstring.

In [8]:
QUESTION_PREFIX = "Question: "
OPTIONS_PREFIX = "Answer Choices: "
CHAIN_OF_THOUGHT_PREFIX = "Rationale: "
ANSWER_PREFIX = "Correct Answer: "
FEWSHOT_SEPARATOR = "\n\n\n"

def make_prompt(*, main_question, fewshot_examples):
  """
  Your goal is to produce the same prompt as the EXAMPLE_3SHOT_CHAIN_OF_THOUGHT automatically

  For each few-shot question, make sure to follow the following rules:
  1. Each question begins with QUESTION_PREFIX, after which you should print the question without leading/traiiling spaces (if any)
  2. After the question, provide space-separated options. Each option should be put in double brackets, followed by option text, e.g. "(A) 146%"
  3. Then, provide the answer as a single letter (A-E)
  4. Finally, add trailing newlines from FEWSHOT_SEPARATOR

  Your final prompt should contain all fewshot_examples (in order), separated with FEWSHOT_SEPARATOR, then follow with main_question.
  The main_question should contain the question and options formatted the same way as in FEWSHOT_EXAMPLES.
  After that, you should prompt the model to produce an explanation (rationale) for the answer.

  Please make sure your prompt contains no leading/trailing newlines or spaces, same as in EXAMPLE_3SHOT_CHAIN_OF_THOUGHT
  """

  prompt = ''
  for f_s in fewshot_examples:
    prompt += f"{QUESTION_PREFIX}{f_s['question']}\n"
    prompt += f"{OPTIONS_PREFIX}({' ('.join(f_s['options']).replace(')', ') ')}\n"
    prompt += f"{CHAIN_OF_THOUGHT_PREFIX}{f_s['rationale']}\n"
    prompt += f"{ANSWER_PREFIX}{f_s['correct']}{FEWSHOT_SEPARATOR}"

  prompt += f"{QUESTION_PREFIX}{main_question['question']}\n"
  prompt += f"{OPTIONS_PREFIX}({' ('.join(main_question['options']).replace(')', ') ')}\n"
  prompt += f"{CHAIN_OF_THOUGHT_PREFIX}"

  return prompt.strip()



generated_fewshot_prompt = make_prompt(main_question=data[150], fewshot_examples=(data[30], data[20], data[5]))
assert generated_fewshot_prompt == EXAMPLE_3SHOT_CHAIN_OF_THOUGHT, "prompts don't match"
assert generated_fewshot_prompt != make_prompt(main_question=data[150], fewshot_examples=())
assert generated_fewshot_prompt.endswith(make_prompt(main_question=data[150], fewshot_examples=()))

print("Well done!")

# Hint: if two prompts do not match, you may find it usefull to use https://www.diffchecker.com or similar to find the difference

NameError: name 'EXAMPLE_3SHOT_CHAIN_OF_THOUGHT' is not defined

__Task 7 (1 points):__ Evaluate your prompt.

Please run the model on the entire dataset and measure it's accuracy.
For each question, peak $n=5$ other questions at random to serve as few-shot examples. Make sure not to accidentally sample the main_question among few-shot examples. For scientific evaluation, it is also a good practice to split the data into two parts: one for eval, and another for few-shot examples. However, doing so is optional in this homework.

The tricky part is when to stop generating: if you don't control for this, your model can accidentally generate a whole new question - and promptyly answer it :) To make sure you get the correct answer, stop generating tokens when the model is done explaining it's solution. To circumvent this, you need to __stop generating as soon as the model generates Final Answer: [A-E]__
To do so, you can either generate manually (see low-level generation above) or use [transformers stopping criteria](https://discuss.huggingface.co/t/implimentation-of-stopping-criteria-list/20040/2), whichever you prefer.

If you do everything right, the model should be much better than random. However, please __do not expect miracles__: this is far from the best models, and it will perform much worse than an average human.

In [6]:
NUM_SAMPLES = 0    # use this to count how many samples you evaluated
NUM_RESPONDED = 0  # how many times did the model produce Correct Answer: (letter) in it's response. use as a sanity check.
NUM_CORRECT = 0    # how many times did the model's chosen answer (letter) match the correct answer

In [ ]:
< A whole lot of your code here >

# Optionally, consider inferencing multiple sentences in a batch for faster inference;
# If you choose to batch outputs, make sure the results are the same as with batch=1 (using greedy inference)

In [16]:
import numpy as np
from sklearn.model_selection import train_test_split
from transformers import StoppingCriteriaList, StoppingCriteria
from tqdm import tqdm

train_few_shot, test_main_question = train_test_split(data, test_size=0.165, random_state=2023)

# Возникали ошибки с cuda, поэтому возьмём вот это решение из предложенной ссылочки
# https://discuss.huggingface.co/t/implimentation-of-stopping-criteria-list/20040/7
class StoppingCriteriaSub(StoppingCriteria):

    def __init__(self, stops = [], encounters=1):
        super().__init__()
        self.stops = [stop.to("cuda") for stop in stops]

    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor):
        for stop in self.stops:
            if torch.all((stop == input_ids[0][-len(stop):])).item():
                return True

        return False


stop_words = [": A", ": B", ": C", ": D", ": E"]
stop_words_ids = [tokenizer(stop_word, return_tensors='pt')['input_ids'].squeeze() for stop_word in stop_words]
stopping_criteria = StoppingCriteriaList([StoppingCriteriaSub(stops=stop_words_ids)])

In [ ]:
NUM_SAMPLES = 0    # use this to count how many samples you evaluated
NUM_RESPONDED = 0  # how many times did the model produce Correct Answer: (letter) in it's response. use as a sanity check.
NUM_CORRECT = 0    # how many times did the model's chosen answer (letter) match the correct answer

correct_answers, generated_answers, generated = [], [], []
with tqdm(total=len(test_main_question)) as pbar:
  for i in range(len(test_main_question)):
      correct_answer = test_main_question[i]['correct']
      correct_answers.append(correct_answer)

      few_shots = np.random.choice(train_few_shot, 5)
      prompt = make_prompt(main_question=test_main_question[i],
                          fewshot_examples=few_shots)

      batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
      torch.manual_seed(2023)
      output_tokens = model.generate(**batch,
                                    max_new_tokens=1024, do_sample=True,
                                    stopping_criteria=stopping_criteria)
      generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())
      generated.append(generated_answer)
      generated_answers.append(generated_answer[-1])

      if generated_answer[-1] == test_main_question[i]['correct']:
          NUM_CORRECT += 1
      if generated_answer[-17: -1] == "Correct Answer: ":
          NUM_RESPONDED += 1
      NUM_SAMPLES += 1
      pbar.update(1)


In [ ]:
print("Responded : %.4f" % NUM_RESPONDED / NUM_SAMPLES)
print("Accuracy (when responded): %.4f" % NUM_CORRECT / NUM_RESPONDED)
print("Accuracy (overall): %.4f" % NUM_CORRECT / NUM_SAMPLES)

if NUM_RESPONDED / NUM_SAMPLES < 0.9:
  print("Something is wrong with the evaluation technique (for 5-shot CoT): the model refuses to answer too many questions.")
  print("Make sure you generate enough tokens that the model can produce a correct answer.")
  print("When in doubt, take a look at the full model output. You can often spot errors there.")

### Решение проблемы с плохим обучением

Увидел, что у меня не работает модель. Проверил на конкретном примере и понял, что напрасно ждал час, пока обучится, потому что критерий остановки генерации ответа не сработал. 

Почитал вот [эту проблемку на гитхабе](https://github.com/huggingface/transformers/issues/22436?ysclid=lpq92y6gyr436024385) и понял, что надо поменять модель.

In [17]:
torch.cuda.empty_cache()

In [3]:
model_name = 'tiiuae/falcon-7b'

# loading Llama tokenizer ...
tokenizer = transformers.AutoTokenizer.from_pretrained(model_name, device_map=device)
tokenizer.pad_token_id = tokenizer.eos_token_id

# ... and the model itself
model = transformers.AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map='auto',
    torch_dtype=torch.float16,
    low_cpu_mem_usage=True,
    offload_state_dict=True
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [14]:
few_shots = np.random.choice(train_few_shot, 5)
prompt = make_prompt(main_question=test_main_question[0],
                  fewshot_examples=few_shots)

batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
torch.manual_seed(2023)
output_tokens = model.generate(**batch, 
                            max_new_tokens=1024, do_sample=True, 
                            stopping_criteria=stopping_criteria,
                            pad_token_id=tokenizer.eos_token_id)
generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())
print(generated_answer)

Generating stopped
 600(0.8) + 80 = 480 golf balls
640 = 480
R = 1.6
640(0.6) - 80 = 480
Answer E.
Correct Answer: E


In [18]:
NUM_SAMPLES = 0    # use this to count how many samples you evaluated
NUM_RESPONDED = 0  # how many times did the model produce Correct Answer: (letter) in it's response. use as a sanity check.
NUM_CORRECT = 0    # how many times did the model's chosen answer (letter) match the correct answer

correct_answers, generated_answers, generated = [], [], []
with tqdm(total=len(test_main_question)) as pbar:
  for i in range(len(test_main_question)):
      correct_answer = test_main_question[i]['correct']
      correct_answers.append(correct_answer)

      few_shots = np.random.choice(train_few_shot, 5)
      prompt = make_prompt(main_question=test_main_question[i],
                          fewshot_examples=few_shots)

      batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
      torch.manual_seed(2023)
      output_tokens = model.generate(**batch,
                                    max_new_tokens=1024, do_sample=True,
                                    stopping_criteria=stopping_criteria,
                                    pad_token_id=tokenizer.eos_token_id)

      generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())
      generated.append(generated_answer)
      generated_answers.append(generated_answer[-1])

      if generated_answer[-1] == test_main_question[i]['correct']:
          NUM_CORRECT += 1
      if generated_answer[-17: -1] == "Correct Answer: ":
          NUM_RESPONDED += 1
      NUM_SAMPLES += 1
      pbar.update(1)

100%|██████████| 42/42 [11:40<00:00, 16.67s/it]


In [21]:
print("Responded : %.4f" % (NUM_RESPONDED / NUM_SAMPLES))
print("Accuracy (when responded): %.4f" % (NUM_CORRECT / NUM_RESPONDED))
print("Accuracy (overall): %.4f" % (NUM_CORRECT / NUM_SAMPLES))

if NUM_RESPONDED / NUM_SAMPLES < 0.9:
  print("Something is wrong with the evaluation technique (for 5-shot CoT): the model refuses to answer too many questions.")
  print("Make sure you generate enough tokens that the model can produce a correct answer.")
  print("When in doubt, take a look at the full model output. You can often spot errors there.")

Responded : 0.6429
Accuracy (when responded): 0.1481
Accuracy (overall): 0.0952
Something is wrong with the evaluation technique (for 5-shot CoT): the model refuses to answer too many questions.
Make sure you generate enough tokens that the model can produce a correct answer.
When in doubt, take a look at the full model output. You can often spot errors there.


В целом, модель приемлемо обучилась, выдавая правильный ответ в 15% случаев. Восстание машин ещё не близко! По крайней мере, если они обучены мной.

In [22]:
# хотел полечить проблему NUM_RESPONDED / NUM_SAMPLES < 0.9, но оч долго работает
NUM_SAMPLES = 0    # use this to count how many samples you evaluated
NUM_RESPONDED = 0  # how many times did the model produce Correct Answer: (letter) in it's response. use as a sanity check.
NUM_CORRECT = 0    # how many times did the model's chosen answer (letter) match the correct answer

correct_answers, generated_answers, generated = [], [], []
with tqdm(total=len(test_main_question)) as pbar:
  for i in range(len(test_main_question)):
      correct_answer = test_main_question[i]['correct']
      correct_answers.append(correct_answer)

      few_shots = np.random.choice(train_few_shot, 5)
      prompt = make_prompt(main_question=test_main_question[i],
                          fewshot_examples=few_shots)

      batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
      torch.manual_seed(2023)
      output_tokens = model.generate(**batch,
                                    max_new_tokens=1024, do_sample=True,
                                    num_beams=4,
                                    stopping_criteria=stopping_criteria,
                                    pad_token_id=tokenizer.eos_token_id)

      generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())
      generated.append(generated_answer)
      generated_answers.append(generated_answer[-1])

      if generated_answer[-1] == test_main_question[i]['correct']:
          NUM_CORRECT += 1
      if generated_answer[-17: -1] == "Correct Answer: ":
          NUM_RESPONDED += 1
      NUM_SAMPLES += 1
      pbar.update(1)

  0%|          | 0/42 [01:48<?, ?it/s]


KeyboardInterrupt: 

__Task 8 (2 points)__ Experiment time!
<img width=200px src=https://www.evolvefish.com/cdn-cgi/image/quality%3D85/assets/images/Apparel/TShirtsWomenCont/Main/EF-APP-CWT-00068(Main).jpg>

Your final quest is to use the testbench you've just written to answer one of the following questions:

### Option 1: How many shots do you need?

How does model accuracy change with the number of fewshot examples?

a. check if the model accuracy changes as you increase/decrease the number of "shots"

b. try to prompt-engineer a model into giving the best rationale __without__ any few-shot examples, i.e. zero-shot

For zero-shot mode, feel free to use wild prompt-engineering or modify the inference procedure.

### Option 2: Is this prompting tecnique reliable?

_Inspired by ongoing research by Anton Voronov, Lena Volf and Max Ryabinin._

For this option, you need to check if the model behavior (and hence, accuracy) is robust to perturbations in the input prompt.

a. Does the accuracy degrade if you provide wrong answers to few-shot examples? (make sure to modify rationale if it contains answer in the end)

b. Does it degrade if you replace question/answer prompts with "Q" and "A"? What if you write both on the same line? Change few-shot separators?



### Option 3: Inference Matters

There are many ways to inference the model, not all of them equal.

a. check whether greedy inference or beam search affects model generation quality

b. implement and evaluate sampling with voting (see explanation below).


The voting technique(b) should work as follows: first, you generate k (e.g. 50) "attempts" at an answer using nucleus sampling (or a similar technique).
Then, you count how many of those attempts chose a particular option (A, B, etc) as the final answer. The option that was chosen most frequently has the most "votes", and therefore "wins".

To speed up voting, you may want to generate these attempts in parallel as a batch. That should be very easy to implement: just run `model.generate` on a list with multiple copies of the same prompt.




================================================

__Common rules:__ You will need to test both hypothes (A and B) in the chosen option. You may choose to replace one of them with your own idea - but please ask course staff in advance (via telegram) if you want full points.

Feel free to organize your code and report as you see fit - but please make sure it's readable and the code runs top-to-bottom :)
Write a short informal report about what you tried and, in doing so, what did you found. Minimum of 2 paragraphs; more is ok; creative visualizations are welcome.

You are allowed (but not required) to prompt the model into generating a report for you --- or helping you write one. However, if you do so, make sure that it is still human-readable :)



In [ ]:
# feel free to organize your solution as you see fit

# Option 1

Намого более простой в реализации с точки зрения нового кода кажется опция 1, её и выберу

## A

In [24]:
sizes = [1, 10]
results = {1:{'NUM_CORRECT': 0,
              'NUM_RESPONDED': 0,
              'NUM_SAMPLES': 0},
           10:{'NUM_CORRECT': 0,
              'NUM_RESPONDED': 0,
              'NUM_SAMPLES': 0}}

# провожу эксперимент на усечённом датасете, поскольку хочу спать
# отмена, понижение числа токенов ускоряет работу

with tqdm(total=len(test_main_question)) as pbar:
  for size in sizes:
    for i in range(len(test_main_question)):
        correct_answer = test_main_question[i]['correct']

        few_shots = np.random.choice(train_few_shot, size)
        prompt = make_prompt(main_question=test_main_question[i],
                            fewshot_examples=few_shots)

        batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
        torch.manual_seed(2023)
        output_tokens = model.generate(**batch,
                                    max_new_tokens=1024, do_sample=True,
                                    stopping_criteria=stopping_criteria,
                                    pad_token_id=tokenizer.eos_token_id)
        generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())

        if generated_answer[-1] == correct_answer:
            results[size]['NUM_CORRECT'] += 1
        if generated_answer[-17: -1] == "Correct Answer: ":
            results[size]['NUM_RESPONDED'] += 1
        results[size]['NUM_SAMPLES'] += 1
        pbar.update(1)

43it [11:00, 25.22s/it]                        This is a friendly reminder - the current text generation call will exceed the model's predefined maximum length (2048). Depending on the model, you may observe exceptions, performance degradation, or nothing at all.
48it [13:12, 27.26s/it]Token indices sequence length is longer than the specified maximum sequence length for this model (2135 > 2048). Running this sequence through the model will result in indexing errors
54it [14:30, 16.13s/it]


KeyboardInterrupt: 

In [26]:
results[7] = {'NUM_CORRECT': 0,
              'NUM_RESPONDED': 0,
              'NUM_SAMPLES': 0}

for i in range(len(test_main_question)):
    correct_answer = test_main_question[i]['correct']

    few_shots = np.random.choice(train_few_shot, 7)
    prompt = make_prompt(main_question=test_main_question[i],
                        fewshot_examples=few_shots)

    batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    torch.manual_seed(2023)
    output_tokens = model.generate(**batch,
                                max_new_tokens=512, do_sample=True,
                                stopping_criteria=stopping_criteria,
                                pad_token_id=tokenizer.eos_token_id)
    generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())

    if generated_answer[-1] == correct_answer:
        results[7]['NUM_CORRECT'] += 1
    if generated_answer[-17: -1] == "Correct Answer: ":
        results[7]['NUM_RESPONDED'] += 1
    results[7]['NUM_SAMPLES'] += 1
    pbar.update(1)

In [27]:
results

{1: {'NUM_CORRECT': 10, 'NUM_RESPONDED': 21, 'NUM_SAMPLES': 42},
 10: {'NUM_CORRECT': 3, 'NUM_RESPONDED': 7, 'NUM_SAMPLES': 12},
 7: {'NUM_CORRECT': 6, 'NUM_RESPONDED': 26, 'NUM_SAMPLES': 42}}

## B

In [31]:
QUESTION_PREFIX = "Question: "
OPTIONS_PREFIX = "Answer Choices: "
CHAIN_OF_THOUGHT_PREFIX = "Rationale: "
ANSWER_PREFIX = "Correct Answer: "
FEWSHOT_SEPARATOR = "\n\n\n"

def make_zero_shot_prompt(*, main_question):

  prompt = "Answer the question using Math. You are to choose the correct answer in format 'Correct Answer: *letter choice*'\n"
  prompt += f"{QUESTION_PREFIX}{main_question['question']}\n"
  prompt += f"{OPTIONS_PREFIX}({' ('.join(main_question['options']).replace(')', ') ')}\n"
  prompt += f"{CHAIN_OF_THOUGHT_PREFIX}"

  return prompt.strip()

In [32]:
print(make_zero_shot_prompt(main_question=data[150]))

Answer the question using Math. You are to choose the correct answer in format 'Correct Answer: *letter choice*'
Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Rationale:


In [33]:
results[0] = {'NUM_CORRECT': 0,
              'NUM_RESPONDED': 0,
              'NUM_SAMPLES': 0}

# провожу эксперимент на усечённом датасете, поскольку хочу спать
# отмена, понижение числа токенов ускоряет работу

with tqdm(total=len(test_main_question)) as pbar:
  for i in range(len(test_main_question)):
      correct_answer = test_main_question[i]['correct']

      prompt = make_zero_shot_prompt(main_question=test_main_question[i])
      batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
      torch.manual_seed(2023)
      output_tokens = model.generate(**batch,
                                    max_new_tokens=1024, do_sample=True,
                                    stopping_criteria=stopping_criteria,
                                    pad_token_id=tokenizer.eos_token_id)
      generated_answer = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu())


      if generated_answer[-1] == correct_answer:
          results[size]['NUM_CORRECT'] += 1
      if generated_answer[-17: -1] == "Correct Answer: ":
          results[size]['NUM_RESPONDED'] += 1
      results[size]['NUM_SAMPLES'] += 1
      pbar.update(1)

100%|██████████| 42/42 [14:09<00:00, 20.23s/it]


In [36]:
generated_answer

'\n1) The cost of an _______ _______ is $18. Let _______ _______ be the cost of the _______ _______. The cost of a _______ _______ is 3 times the cost of an _______ _______. Therefore using the logic above, the cost of a _______ _______ is 3 times the cost of an _______ _______.\n2) The cost of an _______ _______ is $18. Let _______ _______ be the cost of an _______ _______. The cost of a _______ _______ is 3 times the cost of an _______ _______. Therefore the cost of an _______ _______ is 3 times the cost of a _______ _______.\n3) The cost of an _______ _______ is $18. Let _______ _______ be the cost of a _______ _______. (A) The cost of an _______ _______ is 6 times the cost of an _______ _______. (B) The cost of an _______ _______ is 3 times the cost of an _______ _______. (C) The cost of an _______ is $10. (D) The cost of an _______ _______ is $15. (E) The cost of a _______ _______ is 9 times that of an _______ _______.. Therefore using the logic above, the cost of a _______ ______

Что-то явно пошло не так.

## Выводы

In [34]:
results[5] = {'NUM_CORRECT': NUM_CORRECT,
              'NUM_RESPONDED': NUM_RESPONDED,
              'NUM_SAMPLES': NUM_SAMPLES}

In [35]:
results

{1: {'NUM_CORRECT': 10, 'NUM_RESPONDED': 21, 'NUM_SAMPLES': 42},
 10: {'NUM_CORRECT': 3, 'NUM_RESPONDED': 7, 'NUM_SAMPLES': 54},
 7: {'NUM_CORRECT': 6, 'NUM_RESPONDED': 26, 'NUM_SAMPLES': 42},
 0: {'NUM_CORRECT': 0, 'NUM_RESPONDED': 0, 'NUM_SAMPLES': 0},
 5: {'NUM_CORRECT': 0, 'NUM_RESPONDED': 0, 'NUM_SAMPLES': 0}}

Вижу, что нашаманил не то в словаре. Очень лень в 6-30 поправлять это кодом и заново прогонять корректно, поэтому быстро вручную скорректируем словарь и сравним полученные результаты:

* 5
Responded : 0.6429
Accuracy (when responded): 0.1481
Accuracy (overall): 0.0952

* 0 получим, вычтя 10 отсюда из 10 в новой версии словаря
{1: {'NUM_CORRECT': 10, 'NUM_RESPONDED': 21, 'NUM_SAMPLES': 42},
 10: {'NUM_CORRECT': 3, 'NUM_RESPONDED': 7, 'NUM_SAMPLES': 12},
 7: {'NUM_CORRECT': 6, 'NUM_RESPONDED': 26, 'NUM_SAMPLES': 42}}

In [37]:
results[0]['NUM_CORRECT'] = 0
results[0]['NUM_RESPONDED'] = 0
results[0]['NUM_SAMPLES'] = 42

results[5]['NUM_CORRECT'] = round(0.0952*42, 0)
results[5]['NUM_RESPONDED'] = round(0.6429*42, 0)
results[5]['NUM_SAMPLES'] = 42

results

{1: {'NUM_CORRECT': 10, 'NUM_RESPONDED': 21, 'NUM_SAMPLES': 42},
 10: {'NUM_CORRECT': 3, 'NUM_RESPONDED': 7, 'NUM_SAMPLES': 54},
 7: {'NUM_CORRECT': 6, 'NUM_RESPONDED': 26, 'NUM_SAMPLES': 42},
 0: {'NUM_CORRECT': 0, 'NUM_RESPONDED': 0, 'NUM_SAMPLES': 42},
 5: {'NUM_CORRECT': 4.0, 'NUM_RESPONDED': 27.0, 'NUM_SAMPLES': 42}}

In [42]:
import pandas as pd

results_df = pd.DataFrame(results).T.drop(index=10)
results_df['perc_correct'] = results_df.NUM_CORRECT / results_df.NUM_SAMPLES * 100
results_df['perc_correct_among_responded'] = results_df.NUM_CORRECT / results_df.NUM_RESPONDED * 100
results_df['perc_responded'] = results_df.NUM_RESPONDED / results_df.NUM_SAMPLES * 100
results_df.sort_values('perc_correct', ascending=False)

NUM_CORRECT  NUM_RESPONDED  NUM_SAMPLES  perc_correct  \
1         10.0           21.0         42.0     23.809524   
7          6.0           26.0         42.0     14.285714   
5          4.0           27.0         42.0      9.523810   
0          0.0            0.0         42.0      0.000000   

   perc_correct_among_responded  perc_responded  
1                     47.619048       50.000000  
7                     23.076923       61.904762  
5                     14.814815       64.285714  
0                           NaN        0.000000

Как ни странно, модел лучше всего обучилась на 1-shot режиме. Пример с 7-shot мог бы выступить сильнее, но я сам чуть-чуть подпортил эксперимент, снизив в 2 раза количество `max_new_tokens`.